### DISCUSSION TITLE
[Question]: Is there a way to call actions *before* calling the LLM?

### DISCUSSION THREAD

**Discussion by dividor** [danny-avila/LibreChat#2577] [Q&A] [answered]
### What is your question?

Is it possible to make an API call on every user input, where if there is an image or text response, display it an not call the LLM?

### More Details

We have a server which has long-term memories, basically an API end point which when called with chat history can return a previous result. This result may be an image, or some text, for a similar questions asked by any user. It is external to Librachat's cache, a resource with its own management user interface.

We can set it up as an action so the LLM calls it. This works well if the answer is text, as the result is processed by the LLM, but if the result is a base64 encoded image, it fails. It's also a bit slow, as the LLM has to decide to call an action, call it, then process the results.

What we would love to be able to do is have the long-term memory called for *every* input before calling the LLM. If there is a match, then have it displayed (image or text), without calling the LLM.

Might this be possible please, if so, what is the suggested approach?

Thanks! 

### What is the main subject of your question?

Other

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by dividor** [comment]
I think the solution is to create a plugin which calls my API. I have found the documentation for this [here](https://github.com/danny-avila/LibreChat/blob/main/docs/features/plugins/make_your_own.md), but on running the plugins endpoint in LibreChat, I instead see a plugin store - which is super cool - but It's not clear how the instructions work with this. For example, I see a plugin called 'Earth' which if I install works perfectly. But I cannot see any files when I exec into LibreChat's docker container. I see plugin files, but not all of the plugins in the store.

Are there perhaps some updated instructions I can follow to create a plugin?

**Reply by dividor** [reply, marked as answer]
Ok, I've read the great plugin documentation in more detail. Will try that. Closing this question.

**Comment by danny-avila** [comment]
I mean for a specific action like this, you can code something in and make a non-blocking async call based on the input before it even gets to the tool framework, but this is totally out of the current framework.
